File: fieldwatch 2/fieldwatch/shared/src/commonMain/kotlin/com/fieldwatch/shared/outbreak/OutbreakDetector.kt
package com.fieldwatch.shared.outbreak

import com.fieldwatch.shared.geo.GeoUtils
import com.fieldwatch.shared.model.DiseaseReport
import com.fieldwatch.shared.model.OutbreakCluster
import com.fieldwatch.shared.model.ReportStatus
import com.fieldwatch.shared.model.SeverityLevel

class OutbreakDetector(
    private val clusterRadiusKm: Double = 25.0,
    private val timeWindowDays: Int = 10,
    private val alertThresholdReports: Int = 3
) {

    /**
     * Scans reports to detect outbreaks based on spatiotemporal clustering.
     */
    fun detectOutbreaks(reports: List<DiseaseReport>): List<OutbreakCluster> {
        val now = System.currentTimeMillis()
        val windowMillis = timeWindowDays.toLong() * 24 * 60 * 60 * 1000

        // Filter active reports within the sliding time window, excluding resolved/rejected
        val recentReports = reports.filter { report ->
            report.status != ReportStatus.REJECTED &&
            report.status != ReportStatus.RESOLVED &&
            (now - report.timestampMillis) <= windowMillis
        }

        // Group by disease
        val byDisease = recentReports.groupBy { it.suspectDiseaseId ?: "UNKNOWN" }
        val clusters = mutableListOf<OutbreakCluster>()

        for ((diseaseId, diseaseReports) in byDisease) {
            if (diseaseReports.size < alertThresholdReports) continue

            // Spatial clustering (DBSCAN-like single linkage)
            val visited = mutableSetOf<String>()

            for (report in diseaseReports) {
                if (report.id in visited) continue

                val clusterMembers = mutableListOf(report)
                visited.add(report.id)

                val queue = mutableListOf(report)
                while (queue.isNotEmpty()) {
                    val current = queue.removeAt(0)
                    for (candidate in diseaseReports) {
                        if (candidate.id !in visited) {
                            val dist = GeoUtils.distanceBetween(current.location, candidate.location)
                            if (dist <= clusterRadiusKm) {
                                visited.add(candidate.id)
                                clusterMembers.add(candidate)
                                queue.add(candidate)
                            }
                        }
                    }
                }

                if (clusterMembers.size >= alertThresholdReports) {
                    val locations = clusterMembers.map { it.location }
                    val centroid = GeoUtils.calculateCentroid(locations)
                    val totalAffected = clusterMembers.sumOf { it.affectedCount }
                    val totalMortality = clusterMembers.sumOf { it.mortalityCount }

                    // Determine max distance from centroid as bounding cluster radius
                    val maxDist = clusterMembers.maxOfOrNull {
                        GeoUtils.distanceBetween(centroid, it.location)
                    } ?: clusterRadiusKm
                    val effectiveRadius = maxOf(maxDist + 5.0, 15.0)

                    val severity = when {
                        totalMortality >= 10 || clusterMembers.size >= 8 -> SeverityLevel.CRITICAL
                        totalMortality >= 3 || clusterMembers.size >= 5 -> SeverityLevel.HIGH
                        clusterMembers.size >= 3 -> SeverityLevel.MODERATE
                        else -> SeverityLevel.LOW
                    }

                    clusters.add(
                        OutbreakCluster(
                            id = "OUTBREAK-${diseaseId.take(4)}-${clusters.size + 1}",
                            diseaseId = diseaseId,
                            diseaseName = diseaseId.replace("_", " ").capitalizeWords(),
                            center = centroid,
                            radiusKm = effectiveRadius,
                            reportIds = clusterMembers.map { it.id },
                            totalAffected = totalAffected,
                            totalMortality = totalMortality,
                            firstReportTime = clusterMembers.minOf { it.timestampMillis },
                            latestReportTime = clusterMembers.maxOf { it.timestampMillis },
                            severity = severity
                        )
                    )
                }
            }
        }

        return clusters
    }

    private fun String.capitalizeWords(): String =
        split(" ").joinToString(" ") { word -> word.lowercase().replaceFirstChar { it.uppercase() } }
}

File: fieldwatch 2/src/utils/outbreakDetector.ts
import { DiseaseReport, OutbreakCluster, SeverityLevel } from '../types';
import { calculateCentroid, distanceBetween } from './geo';

export interface DetectorConfig {
  clusterRadiusKm: number;
  timeWindowDays: number;
  alertThresholdReports: number;
}

export const DEFAULT_DETECTOR_CONFIG: DetectorConfig = {
  clusterRadiusKm: 25,
  timeWindowDays: 14,
  alertThresholdReports: 3,
};

export class OutbreakDetector {
  constructor(private config: DetectorConfig = DEFAULT_DETECTOR_CONFIG) {}

  /**
   * Spatiotemporal clustering algorithm (DBSCAN single-linkage over geographic distance + time window).
   */
  detectOutbreaks(
    reports: DiseaseReport[],
    customConfig?: Partial<DetectorConfig>
  ): OutbreakCluster[] {
    const cfg = { ...this.config, ...customConfig };
    const now = Date.now();
    const windowMillis = cfg.timeWindowDays * 24 * 3600 * 1000;

    // Filter active reports in the window, ignoring resolved/rejected
    const activeReports = reports.filter(
      (r) =>
        r.status !== 'REJECTED' &&
        r.status !== 'RESOLVED' &&
        now - r.timestampMillis <= windowMillis
    );

    // Group by suspect disease
    const byDisease = new Map<string, DiseaseReport[]>();
    for (const report of activeReports) {
      const dId = report.suspectDiseaseId || 'UNKNOWN';
      const existing = byDisease.get(dId) || [];
      existing.push(report);
      byDisease.set(dId, existing);
    }

    const clusters: OutbreakCluster[] = [];

    for (const [diseaseId, diseaseReports] of byDisease.entries()) {
      if (diseaseReports.length < cfg.alertThresholdReports) continue;

      const visited = new Set<string>();

      for (const report of diseaseReports) {
        if (visited.has(report.id)) continue;

        const clusterMembers: DiseaseReport[] = [report];
        visited.add(report.id);

        const queue: DiseaseReport[] = [report];

        while (queue.length > 0) {
          const current = queue.shift()!;
          for (const candidate of diseaseReports) {
            if (!visited.has(candidate.id)) {
              const dist = distanceBetween(current.location, candidate.location);
              if (dist <= cfg.clusterRadiusKm) {
                visited.add(candidate.id);
                clusterMembers.push(candidate);
                queue.push(candidate);
              }
            }
          }
        }

        if (clusterMembers.length >= cfg.alertThresholdReports) {
          const locations = clusterMembers.map((r) => r.location);
          const centroid = calculateCentroid(locations);
          const totalAffected = clusterMembers.reduce((sum, r) => sum + r.affectedCount, 0);
          const totalMortality = clusterMembers.reduce((sum, r) => sum + r.mortalityCount, 0);

          let maxDist = 0;
          for (const m of clusterMembers) {
            const d = distanceBetween(centroid, m.location);
            if (d > maxDist) maxDist = d;
          }
          const effectiveRadius = Math.max(Math.round(maxDist + 5), 15);

          let severity: SeverityLevel = 'MODERATE';
          if (totalMortality >= 10 || clusterMembers.length >= 6) {
            severity = 'CRITICAL';
          } else if (totalMortality >= 3 || clusterMembers.length >= 4) {
            severity = 'HIGH';
          } else if (clusterMembers.length >= 3) {
            severity = 'MODERATE';
          }

          const diseaseName =
            diseaseId === 'fmd'
              ? 'Foot-and-Mouth Disease (FMD)'
              : diseaseId === 'asf'
              ? 'African Swine Fever (ASF)'
              : diseaseId === 'newcastle'
              ? 'Newcastle Disease / Avian Flu'
              : diseaseId === 'fall_armyworm'
              ? 'Fall Armyworm (FAW)'
              : diseaseId === 'cassava_mosaic'
              ? 'Cassava Mosaic Disease (CMD)'
              : diseaseId.toUpperCase();

          clusters.push({
            id: `OUTBREAK-${diseaseId.toUpperCase()}-${clusters.length + 1}`,
            diseaseId,
            diseaseName,
            center: centroid,
            radiusKm: effectiveRadius,
            reportIds: clusterMembers.map((r) => r.id),
            totalAffected,
            totalMortality,
            firstReportTime: Math.min(...clusterMembers.map((r) => r.timestampMillis)),
            latestReportTime: Math.max(...clusterMembers.map((r) => r.timestampMillis)),
            severity,
            status: 'ACTIVE',
          });
        }
      }
    }

    return clusters;
  }
}
